core/src/main/java/com/graphhopper/routing/util/AbstractFlagEncoder.java

```diff
@@ -222,20 +223,35 @@ public double getMaxSpeed() {
     }
 
     /**
-     * @return -1 if no maxspeed found
+     * @return {@link Double#NaN} if no maxspeed found
      */
     protected double getMaxSpeed(ReaderWay way) {
-        double maxSpeed = parseSpeed(way.getTag("maxspeed"));
-        double fwdSpeed = parseSpeed(way.getTag("maxspeed:forward"));
-        if (fwdSpeed >= 0 && (maxSpeed < 0 || fwdSpeed < maxSpeed))
+        double maxSpeed = OSMValueExtractor.stringToKmh(way.getTag("maxspeed"));
+        double fwdSpeed = OSMValueExtractor.stringToKmh(way.getTag("maxspeed:forward"));
+        if (isValidSpeed(fwdSpeed) && (!isValidSpeed(maxSpeed) || fwdSpeed < maxSpeed))
             maxSpeed = fwdSpeed;
 
-        double backSpeed = parseSpeed(way.getTag("maxspeed:backward"));
-        if (backSpeed >= 0 && (maxSpeed < 0 || backSpeed < maxSpeed))
+        double backSpeed = OSMValueExtractor.stringToKmh(way.getTag("maxspeed:backward"));
+        if (isValidSpeed(backSpeed) && (!isValidSpeed(maxSpeed) || backSpeed < maxSpeed))
             maxSpeed = backSpeed;
 
         return maxSpeed;
     }
+    
+    /**
+     * @return <i>true</i> if the given speed is not {@link Double#NaN} or negative
+     */
+    protected boolean isValidSpeed(double speed) {
+        return !Double.isNaN(speed) && speed >= 0;
+    }
+
+    /**
+     * @return <i>true</i> if the given speed is not {@link Double#NaN} and is
+     *         bigger than zero
+     */
+    protected boolean isPositiveSpeed(double speed) {
+        return !Double.isNaN(speed) && speed > 0;
+    }
```

Comment:
why is this necessary?

Reply:
Yeah i was thinking about the same. Wouldn't it be better to ignore 0 speeds completely?

https://overpass-turbo.eu/s/WwG

Most of those ways have an accompanying access tags which already prohibit routing.

Reply:
Yes, your overpass query is convincing, seems to be a good idea to exclude those ways and assume this is a tagging problem. Or is it used for other purposes (e.g. for other speed tags) too?

Reply:
I couldn't find any reasonable uses of speed 0 for `maxspeed:forward`, `maxspeed:backward` and `maxspeed:motorcycle`. It should be safe to ignore it as a tagging error.

Related: https://github.com/graphhopper/graphhopper/issues/367